server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -498,6 +498,29 @@ public MappingMetaData mappingOrDefault(String mappingType) {
         return mappings.get(MapperService.DEFAULT_MAPPING);
     }
 
+    /**
+     * Resolves a type from a mapping-related request into the type that should be used when
+     * merging and updating mappings.
+     *
+     * If the special `_doc` type is provided, then we replace it with the actual type that is
+     * being used in the mappings. This allows typeless APIs such as 'index' or 'put mappings'
+     * to work against indices with a custom type name.
+     */
+    public String resolveDocumentType(String type) {
```

Comment:
7.x is a bit simpler since this method no longer takes a type parameter. I considered having it as part of `mappingOrDefault` but then it felt wrong to not have this logic in `mapping` as well, while the automatic replacement does not always make sense for call sites of the `mapping` method such as `MapperService#assertMappingVersion`.

Reply:
👍